File: .config/ai-sync/sync/add-project.ts
import { existsSync, readdirSync } from "fs";
import { resolve, join, dirname } from "path";

interface ProjectConfig {
  path: string;
  skills?: string[];
}

interface Config {
  projects: ProjectConfig[];
}

function getSyncDir(): string {
  const isCompiled = import.meta.dir.startsWith("/$bunfs");
  if (isCompiled) {
    return dirname(process.execPath);
  }
  return import.meta.dir;
}

function getAvailableSkills(skillsDir: string): string[] {
  if (!existsSync(skillsDir)) {
    return [];
  }
  return readdirSync(skillsDir, { withFileTypes: true })
    .filter((dirent) => dirent.isDirectory())
    .map((dirent) => dirent.name);
}

export async function addProject(): Promise<void> {
  const syncDir = getSyncDir();
  const rootDir = resolve(syncDir, "..");
  const skillsSourceDir = join(rootDir, "skills");
  const configPath = join(syncDir, "config.json");
  const cwd = process.cwd();

  let config: Config = { projects: [] };
  if (existsSync(configPath)) {
    const text = await Bun.file(configPath).text();
    config = JSON.parse(text);
  }

  const existingProject = config.projects.find(
    (p) => resolve(p.path) === resolve(cwd)
  );
  if (existingProject) {
    console.log(`Project already exists: ${cwd}`);
    console.log(`Skills: ${existingProject.skills?.join(", ") || "none"}`);
    return;
  }

  const availableSkills = getAvailableSkills(skillsSourceDir);
  if (availableSkills.length === 0) {
    console.warn(`No skills found in: ${skillsSourceDir}`);
  }

  const newProject: ProjectConfig = {
    path: cwd,
    skills: availableSkills,
  };

  config.projects.push(newProject);

  await Bun.write(configPath, JSON.stringify(config, null, 2) + "\n");

  console.log(`Added project: ${cwd}`);
  console.log(`Skills: ${availableSkills.join(", ") || "none"}`);
}

File: .config/ai-sync/cmd/sync/add_project.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
)

func getAvailableSkills(skillsDir string) []string {
	entries, err := os.ReadDir(skillsDir)
	if err != nil {
		return nil
	}

	var skills []string
	for _, entry := range entries {
		if entry.IsDir() {
			skills = append(skills, entry.Name())
		}
	}
	return skills
}

func AddProject() error {
	rootDir := GetRootDir()
	skillsSourceDir := filepath.Join(rootDir, "build", "skills")
	configPath := filepath.Join(rootDir, "sync-config.yaml")

	cwd, err := os.Getwd()
	if err != nil {
		return fmt.Errorf("getting current directory: %w", err)
	}

	// Load existing config or create new one
	var config *Config
	if _, err := os.Stat(configPath); os.IsNotExist(err) {
		config = &Config{Projects: []ProjectConfig{}}
	} else {
		config, err = LoadConfig(configPath)
		if err != nil {
			return err
		}
	}

	// Check if project already exists
	for _, p := range config.Projects {
		absPath, _ := filepath.Abs(p.Path)
		absCwd, _ := filepath.Abs(cwd)
		if absPath == absCwd {
			fmt.Printf("Project already exists: %s\n", cwd)
			if p.AiStuff != nil && len(p.AiStuff.Skills) > 0 {
				fmt.Printf("Skills: %v\n", p.AiStuff.Skills)
			} else {
				fmt.Println("Skills: none")
			}
			if len(p.Harnesses) > 0 {
				fmt.Printf("Harnesses: %v\n", p.Harnesses)
			} else {
				fmt.Println("Harnesses: none")
			}
			return nil
		}
	}

	availableSkills := getAvailableSkills(skillsSourceDir)
	if len(availableSkills) == 0 {
		fmt.Printf("Warning: No skills found in: %s\n", skillsSourceDir)
	}

	newProject := ProjectConfig{
		Path: cwd,
		AiStuff: &AiStuff{
			Skills: availableSkills,
		},
		Harnesses: SupportedHarnesses,
	}

	config.Projects = append(config.Projects, newProject)

	if err := SaveConfig(configPath, config); err != nil {
		return err
	}

	fmt.Printf("Added project: %s\n", cwd)
	if len(availableSkills) > 0 {
		fmt.Printf("Skills: %v\n", availableSkills)
	} else {
		fmt.Println("Skills: none")
	}
	fmt.Printf("Harnesses: %v\n", SupportedHarnesses)

	return nil
}
